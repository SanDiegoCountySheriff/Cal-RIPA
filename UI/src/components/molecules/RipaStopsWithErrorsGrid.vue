<template>
  <div class="ripa-stops-with-errors-grid">
    <v-card-title>
      <v-text-field
        v-model="search"
        append-icon="mdi-magnify"
        label="Search"
        single-line
        hide-details
      ></v-text-field>
    </v-card-title>
    <v-data-table
      :headers="headers"
      :items="stopsWithErrors"
      :search="search"
      @click:row="handleRowClick"
      sort-by="id"
      single-select
    >
      <template v-slot:no-data>
        <div>No Data</div>
      </template>
    </v-data-table>
  </div>
</template>

<script>
export default {
  name: 'ripa-stops-with-errors-grid',

  data() {
    return {
      search: '',
      persons: [],
      headers: [
        { text: 'Stop Date', value: 'stopDate', width: '120' },
        { text: 'Stop Time', value: 'stopTime', width: '120' },
        { text: 'Error Code', value: 'errorCode', width: '120' },
        { text: 'Error Text', value: 'errorText' },
      ],
    }
  },

  methods: {
    init() {
      this.stopsWithErrors = this.items.map(item => {
        return {
          internalId: item.internalId || 'Missing Internal Id',
          stopDate: item.apiStop?.date || 'N/A',
          stopTime: item.apiStop.time || 'N/A',
          errorCode: item.statusCode,
          errorText: item.statusError.message
            ? item.statusError.message
            : item.statusError,
        }
      })
    },

    handleRowClick(item, row) {
      row.select(true)
      if (this.onEditStop) {
        this.onEditStop(item.internalId)
      }
    },
  },

  watch: {
    items(val) {
      this.stopsWithErrors = val
    },
  },

  created() {
    this.init()
  },

  props: {
    items: {
      type: Array,
      default: () => [],
    },
    onEditStop: {
      type: Function,
      default: () => {},
    },
  },
}
</script>

<style lang="scss">
.ripa-stops-with-errors-grid {
  .v-data-table__wrapper {
    tr:hover {
      cursor: pointer !important;
    }
  }
}
</style>
