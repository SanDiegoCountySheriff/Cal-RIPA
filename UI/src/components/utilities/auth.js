import { PublicClientApplication } from '@azure/msal-browser'

export const doSignIn = async () => {
  await this.$msalInstance
    .loginPopup({})
    .then(() => {
      const myAccounts = this.$msalInstance.getAllAccounts()
      this.account = myAccounts[0]
      this.$emitter.emit('login', this.account)
    })
    .catch(error => {
      console.error(`error during authentication: ${error}`)
    })
}
