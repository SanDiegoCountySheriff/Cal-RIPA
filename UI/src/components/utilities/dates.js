export const formatDate = dateStr => {
  if (dateStr && dateStr.length > 0) {
    const str = dateStr.replace(/-/g, '')
    const year = str.substring(0, 4)
    const month = str.substring(4, 6)
    const day = str.substring(6, 8)
    return `${year}-${month}-${day}`
  }

  return ''
}
