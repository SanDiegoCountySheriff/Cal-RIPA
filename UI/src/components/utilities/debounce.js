export const debounce = (func, delay, immediate) => {
  let timerId
  return (...args) => {
    const boundFunc = func.bind(this, ...args)
    clearTimeout(timerId)
    if (immediate && !timerId) {
      boundFunc()
    }
    const calleeFunc = immediate
      ? () => {
          timerId = null
        }
      : boundFunc
    timerId = setTimeout(calleeFunc, delay)
  }
}
