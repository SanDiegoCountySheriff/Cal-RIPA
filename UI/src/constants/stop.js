export const SUBMISSION_STATUSES = [
  {
    text: 'Failed',
    value: 'Failed',
    editable: true,
  },
  {
    text: 'Submitted',
    value: 'Submitted',
    editable: false,
  },
  {
    text: 'Unsubmitted',
    value: 'Unsubmitted',
    editable: true,
  },
  {
    text: 'Resubmitted',
    value: 'Resubmitted',
    editable: false,
  },
]
