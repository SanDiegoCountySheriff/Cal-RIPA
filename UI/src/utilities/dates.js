import differenceInHours from 'date-fns/differenceInHours'
import { differenceInYears as diffInYears } from 'date-fns'
import format from 'date-fns/format'
import isValid from 'date-fns/isValid'
import endOfDay from 'date-fns/endOfDay'

export const formatDate = dateStr => {
  if (dateStr && dateStr.length > 0) {
    const str = dateStr.replace(/-/g, '')
    const year = str.substring(0, 4)
    const month = str.substring(4, 6)
    const day = str.substring(6, 8)
    return `${year}-${month}-${day}`
  }

  return ''
}

export const isValidDate = dateStr => {
  const date = parseDate(dateStr, '00:00')
  return isValid(date)
}

const parseDate = (dateStr, timeStr) => {
  return new Date(`${dateStr}T${timeStr}`)
}

export const dateNotInFuture = (dateStr, timeStr) => {
  const date = parseDate(dateStr, timeStr || '00:00')
  return new Date().getTime() >= date.getTime()
}

export const differenceInYears = date => {
  const diff = diffInYears(new Date(), new Date(date))
  return diff
}

export const dateWithinLastHours = (dateStr, timeStr, hours) => {
  const date = parseDate(dateStr, timeStr)
  const diff = differenceInHours(new Date(), date)
  return diff < hours
}

export const formatDateTime = (dateStr, timeStr) => {
  return new Date(`${dateStr}T${timeStr}`)
}

export const formatShortDate = dateStr => {
  const date = parseDate(dateStr, '00:00')
  return format(new Date(date), 'yyyy-MM-dd')
}

export const uniqueId = () => {
  return format(new Date(), 'yyyyMMddkkmm')
}

export const formatToIsoCurrentDate = () => {
  return format(endOfDay(new Date()), `yyyy-MM-dd'T'HH:mm:ss.SSSxxx`)
}

export const formatToIsoDate = date => {
  return format(endOfDay(date), `yyyy-MM-dd'T'HH:mm:ss.SSSxxx`)
}
