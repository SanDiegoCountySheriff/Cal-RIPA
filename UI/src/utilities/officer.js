export const getOfficerStartDate = () => {
  const startDate = localStorage.getItem('ripa_officer_start_date')
  return startDate || null
}

export const getOfficerYearsExperience = () => {
  const yearsExperience = localStorage.getItem('ripa_officer_years_experience')
  return +yearsExperience || null
}

export const getOfficerAssignment = () => {
  const assignment = localStorage.getItem('ripa_officer_assignment')
  return +assignment || null
}

export const getOfficerOtherType = () => {
  const otherType = localStorage.getItem('ripa_officer_other_type')
  return otherType || null
}
