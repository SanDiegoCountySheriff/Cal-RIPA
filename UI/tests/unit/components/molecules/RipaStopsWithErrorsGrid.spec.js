import RipaStopsWithErrorsGrid from '@/components/molecules/RipaStopsWithErrorsGrid.vue'
import { shallowMount } from '@vue/test-utils'
import { defaultStop } from '@/utilities/stop'
import Vuetify from 'vuetify'

describe('Ripa Stops With Errors Grid', () => {
  let vuetify
  let wrapper
  let stop

  beforeEach(() => {
    vuetify = new Vuetify()
    stop = defaultStop()
  })

  afterEach(() => {
    wrapper.destroy()
  })

  const factory = propsData => {
    return shallowMount(RipaStopsWithErrorsGrid, {
      vuetify,
      propsData: {
        ...propsData,
      },
    })
  }

  it('should match snapshot', () => {
    wrapper = factory({ value: stop })

    expect(wrapper.element).toMatchSnapshot()
  })
})
