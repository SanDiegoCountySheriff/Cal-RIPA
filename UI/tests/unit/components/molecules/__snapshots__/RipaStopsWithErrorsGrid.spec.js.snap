// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Ripa Stops With Errors Grid should match snapshot 1`] = `
<div
  class="ripa-stops-with-errors-grid"
  value="[object Object]"
>
  <v-data-table-stub
    customfilter="[Function]"
    customgroup="[Function]"
    customsort="[Function]"
    expanded=""
    expandicon="$expand"
    groupby=""
    groupdesc=""
    headers="[object Object],[object Object]"
    itemclass=""
    itemkey="id"
    items=""
    itemsperpage="10"
    loaderheight="4"
    loadingtext="$vuetify.dataIterator.loadingText"
    locale="en-US"
    mobilebreakpoint="0"
    nodatatext="$vuetify.noDataText"
    noresultstext="$vuetify.dataIterator.noResultsText"
    options="[object Object]"
    page="1"
    selectablekey="isSelectable"
    serveritemslength="-1"
    singleselect="true"
    sortby="id"
    sortdesc=""
    value=""
  />
</div>
`;
